File: lib-segment/src/main/java/websiteschema/mpsegment/hmm/NodeRepository.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package websiteschema.mpsegment.hmm;

import java.io.IOException;
import java.util.*;
import websiteschema.mpsegment.util.ISerialize;
import websiteschema.mpsegment.util.SerializeHandler;

/**
 *
 * @author ray
 */
public class NodeRepository implements ISerialize {

    List<Node> repo = null;
    Map<String, Integer> indexMap = null;

    public NodeRepository() {
        repo = new ArrayList<Node>();
        indexMap = new HashMap<String, Integer>();
    }

    public Node add(Node node) {
        String name = node.getName();
        if (!indexMap.containsKey(name)) {
            int index = repo.size();
            node.setIndex(index);
            repo.add(node);
            indexMap.put(name, index);
            return node;
        } else {
            return repo.get(indexMap.get(name));
        }
    }

    public Node get(String name) {
        if (indexMap.containsKey(name)) {
            int index = indexMap.get(name);
            return repo.get(index);
        } else {
            return null;
        }
    }

    public Node get(int index) {
        if (repo.size() > index) {
            return repo.get(index);
        } else {
            return null;
        }
    }

    public Set<String> keySet() {
        return indexMap.keySet();
    }

    @Override
    public void save(SerializeHandler writeHandler) throws IOException {
        int length = null != repo ? repo.size() : 0;
        writeHandler.serializeInt(length);
        for(int i = 0; i < length; i++) {
            Node node = repo.get(i);
            node.save(writeHandler);
        }
    }

    @Override
    public void load(SerializeHandler readHandler) throws IOException {
        int length = readHandler.deserializeInt();
        if(length > 0) {
            for(int i = 0; i < length; i++) {
                Node node = new Node();
                node.load(readHandler);
                repo.add(node);
                indexMap.put(node.getName(), node.getIndex());
            }
        }
    }
}

File: lib-segment-scala/src/main/scala/websiteschema/mpsegment/hmm/NodeRepository.scala
package websiteschema.mpsegment.hmm

import websiteschema.mpsegment.util.ISerialize
import websiteschema.mpsegment.util.SerializeHandler

class NodeRepository extends ISerialize {

  private var repo = List[Node]()
  private val indexMap = new java.util.HashMap[String, Int]()

  def add(node: Node): Node = {
    val name = node.getName()
    if (!indexMap.containsKey(name)) {
      val index = repo.size
      node.setIndex(index)
      repo = repo :+ node
      indexMap.put(name, index)
      return node
    } else {
      return repo(indexMap.get(name))
    }
  }

  def get(name: String): Node = {
    if (indexMap.containsKey(name)) {
      val index = indexMap.get(name)
      return repo(index)
    } else {
      return null
    }
  }

  def get(index: Int): Node = {
    if (repo.size > index) {
      return repo(index)
    } else {
      return null
    }
  }

  def keySet() = indexMap.keySet


  override def save(writeHandler: SerializeHandler) {
    val length = if (null != repo) repo.size else 0
    writeHandler.serializeInt(length)
    for (i <- 0 until length) {
      val node = repo(i)
      node.save(writeHandler)
    }
  }

  override def load(readHandler: SerializeHandler) {
    val length = readHandler.deserializeInt()
    if (length > 0) {
      for (i <- 0 until length) {
        val node = new Node()
        node.load(readHandler)
        repo = repo :+ node
        indexMap.put(node.getName(), node.getIndex())
      }
    }
  }
}
